benchmark/engine.py

```diff
@@ -0,0 +1,226 @@
+        raise NotImplementedError()
+
+    @classmethod
+    def create(cls, engine_type, project_id):
+        if engine_type is NLUEngines.AMAZON_LEX:
+            return AmazonLex()
+        elif engine_type is NLUEngines.GOOGLE_DIALOGFLOW:
+            return GoogleDialogflow(project_id)
+        elif engine_type is NLUEngines.PICOVOICE_RHINO:
+            return PicovoiceRhino()
+        else:
+            raise ValueError("cannot create %s of type '%s'" %
+                             (cls.__name__, engine_type))
+
+
+class AmazonLex(NLUEngine):
+    def __init__(self):
+        self._client = boto3.client('lex-runtime')
+
+    def process_file(self, path):
+        cache_path = path.replace('.wav', '.amazonlex')
+
+        if os.path.exists(cache_path):
+            with open(cache_path) as f:
+                return json.load(f)
+
+        with open(path, 'rb') as input_audio:
+            response = self._client.post_content(
+                botName='barista',
+                botAlias='$LATEST',
+                userId=str(uuid.uuid4()),
+                contentType='audio/l16; rate=16000; channels=1',
+                accept='text/plain; charset=utf-8',
+                inputStream=input_audio
+            )
+
+        result = dict(intent=response['intentName'],
+                      slots={k: v for k, v in response['slots'].items()
+                             if v is not None},
+                      inputTranscript=response['inputTranscript'])
+
+        with open(cache_path, 'w') as f:
+            json.dump(result, f, indent=2)
+
+        return result
+
+    def __str__(self):
+        return 'Amazon Lex'
+
+
+class GoogleDialogflow(NLUEngine):
+    def __init__(self, project_id):
+        self._project_id = project_id
+
+    def process_file(self, path):
+        cache_path = path.replace('.wav', '.dialogflow')
+
+        if os.path.exists(cache_path):
+            with open(cache_path) as f:
+                return json.load(f)
+
+        session_client = dialogflow.SessionsClient()
+
+        session_id = os.path.basename(path)[0]
+
+        session = session_client.session_path(self._project_id, session_id)
+
+        with open(path, 'rb') as audio_file:
+            input_audio = audio_file.read()
+
+        audio_config = dialogflow.types.InputAudioConfig(
+            audio_encoding=dialogflow.enums.AudioEncoding.AUDIO_ENCODING_LINEAR_16,
+            language_code='en',
+            sample_rate_hertz=16000)
+
+        query_input = dialogflow.types.QueryInput(audio_config=audio_config)
+
+        response = session_client.detect_intent(session=session,
+                                                query_input=query_input,
+                                                input_audio=input_audio)
+
+        result = dict(intent=response.query_result.intent.display_name,
+                      slots=dict())
+
+        for k in ['coffeeDrink', 'size', 'roast', 'numberOfShots', 'sugarAmount', 'milkAmount']:

```

Comment:
do we need to iterate over all keys?

Reply:
I'll change it so that it only iterates over the keys that are in the response.